File: internal/domain/service/catalog_service.go
package service

import (
	"context"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"

	"foto-app/internal/domain/model"
	"foto-app/internal/domain/port"
)

// ---------- Request / Response types ----------

// CreateCourseRequest carries the fields needed to create a course.
type CreateCourseRequest struct {
	Name        string  `json:"name"`
	Description *string `json:"description,omitempty"`
}

// UpdateCourseRequest carries the fields to update a course.
type UpdateCourseRequest struct {
	Name        string  `json:"name"`
	Description *string `json:"description,omitempty"`
	IsActive    *bool   `json:"is_active,omitempty"`
}

// CreateDivisionRequest carries the fields needed to create a division.
type CreateDivisionRequest struct {
	CourseID string `json:"course_id"`
	Name     string `json:"name"`
}

// UpdateDivisionRequest carries the fields to update a division.
type UpdateDivisionRequest struct {
	CourseID string `json:"course_id"`
	Name     string `json:"name"`
	IsActive *bool  `json:"is_active,omitempty"`
}

// CreateBookletRequest carries the fields needed to create a booklet.
type CreateBookletRequest struct {
	CourseID    string          `json:"course_id"`
	DivisionID  string          `json:"division_id"`
	Title       string          `json:"title"`
	Description *string         `json:"description,omitempty"`
	Price       decimal.Decimal `json:"price"`
	Stock       int             `json:"stock"`
	ImageURL    *string         `json:"image_url,omitempty"`
	IsActive    *bool           `json:"is_active,omitempty"`
}

// UpdateBookletRequest carries the fields to update a booklet.
type UpdateBookletRequest struct {
	CourseID    string          `json:"course_id"`
	DivisionID  string          `json:"division_id"`
	Title       string          `json:"title"`
	Description *string         `json:"description,omitempty"`
	Price       decimal.Decimal `json:"price"`
	Stock       int             `json:"stock"`
	ImageURL    *string         `json:"image_url,omitempty"`
	IsActive    *bool           `json:"is_active,omitempty"`
}

// ---------- CatalogService ----------

// CatalogService implements business logic for the catalog domain:
// courses, divisions, and booklets.
type CatalogService struct {
	courseRepo   port.CourseRepository
	divisionRepo port.DivisionRepository
	bookletRepo  port.BookletRepository
}

// NewCatalogService creates a CatalogService with its repository dependencies.
func NewCatalogService(
	courseRepo port.CourseRepository,
	divisionRepo port.DivisionRepository,
	bookletRepo port.BookletRepository,
) *CatalogService {
	return &CatalogService{
		courseRepo:   courseRepo,
		divisionRepo: divisionRepo,
		bookletRepo:  bookletRepo,
	}
}

// ==================== Student-facing methods ====================

// ListActiveBookletsByCourse returns active booklets with stock > 0 for the given course.
func (s *CatalogService) ListActiveBookletsByCourse(ctx context.Context, courseID uuid.UUID) ([]model.Booklet, error) {
	return s.bookletRepo.ListActiveByCourse(ctx, courseID)
}

// ListActiveBooklets returns active booklets with stock > 0, with optional filtering and pagination.
func (s *CatalogService) ListActiveBooklets(ctx context.Context, filter port.BookletFilter) ([]model.Booklet, int, error) {
	filter.AdminView = false
	return s.bookletRepo.List(ctx, filter)
}

// GetBooklet retrieves a single booklet by ID (admin view shows all).
func (s *CatalogService) GetBooklet(ctx context.Context, id uuid.UUID) (*model.Booklet, error) {
	return s.bookletRepo.FindByID(ctx, id)
}

// ListCourses returns all courses.
func (s *CatalogService) ListCourses(ctx context.Context) ([]model.Course, error) {
	return s.courseRepo.List(ctx)
}

// ListDivisionsByCourse returns all divisions for a given course.
func (s *CatalogService) ListDivisionsByCourse(ctx context.Context, courseID uuid.UUID) ([]model.Division, error) {
	// Verify course exists first
	if _, err := s.courseRepo.FindByID(ctx, courseID); err != nil {
		return nil, err
	}
	return s.divisionRepo.ListByCourse(ctx, courseID)
}

// GetCourse retrieves a single course.
func (s *CatalogService) GetCourse(ctx context.Context, id uuid.UUID) (*model.Course, error) {
	return s.courseRepo.FindByID(ctx, id)
}

// ==================== Admin methods — Courses ====================

// CreateCourse creates a new course.
func (s *CatalogService) CreateCourse(ctx context.Context, req CreateCourseRequest) (*model.Course, error) {
	course := &model.Course{
		ID:          uuid.New(),
		Name:        req.Name,
		Description: req.Description,
		IsActive:    true,
	}
	if err := s.courseRepo.Create(ctx, course); err != nil {
		return nil, err
	}
	// Re-fetch to get server-set timestamps
	return s.courseRepo.FindByID(ctx, course.ID)
}

// UpdateCourse updates an existing course.
func (s *CatalogService) UpdateCourse(ctx context.Context, id uuid.UUID, req UpdateCourseRequest) (*model.Course, error) {
	course, err := s.courseRepo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	course.Name = req.Name
	if req.Description != nil {
		course.Description = req.Description
	}
	if req.IsActive != nil {
		course.IsActive = *req.IsActive
	}
	if err := s.courseRepo.Update(ctx, course); err != nil {
		return nil, err
	}
	return s.courseRepo.FindByID(ctx, id)
}

// DeleteCourse deletes a course by ID.
func (s *CatalogService) DeleteCourse(ctx context.Context, id uuid.UUID) error {
	return s.courseRepo.Delete(ctx, id)
}

// ==================== Admin methods — Divisions ====================

// CreateDivision creates a new division.
func (s *CatalogService) CreateDivision(ctx context.Context, req CreateDivisionRequest) (*model.Division, error) {
	courseID, err := uuid.Parse(req.CourseID)
	if err != nil {
		return nil, model.ErrValidation
	}
	// Verify course exists
	if _, err := s.courseRepo.FindByID(ctx, courseID); err != nil {
		return nil, err
	}

	division := &model.Division{
		ID:       uuid.New(),
		CourseID: courseID,
		Name:     req.Name,
		IsActive: true,
	}
	if err := s.divisionRepo.Create(ctx, division); err != nil {
		return nil, err
	}
	return s.divisionRepo.FindByID(ctx, division.ID)
}

// UpdateDivision updates an existing division.
func (s *CatalogService) UpdateDivision(ctx context.Context, id uuid.UUID, req UpdateDivisionRequest) (*model.Division, error) {
	courseID, err := uuid.Parse(req.CourseID)
	if err != nil {
		return nil, model.ErrValidation
	}
	division, err := s.divisionRepo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	division.CourseID = courseID
	division.Name = req.Name
	if req.IsActive != nil {
		division.IsActive = *req.IsActive
	}
	if err := s.divisionRepo.Update(ctx, division); err != nil {
		return nil, err
	}
	return s.divisionRepo.FindByID(ctx, id)
}

// DeleteDivision deletes a division by ID.
func (s *CatalogService) DeleteDivision(ctx context.Context, id uuid.UUID) error {
	return s.divisionRepo.Delete(ctx, id)
}

// ==================== Admin methods — Booklets ====================

// CreateBooklet creates a new booklet.
func (s *CatalogService) CreateBooklet(ctx context.Context, req CreateBookletRequest) (*model.Booklet, error) {
	if req.Price.IsNegative() {
		return nil, model.ErrNegativePrice
	}
	if req.Stock < 0 {
		return nil, model.ErrNegativeStock
	}

	courseID, err := uuid.Parse(req.CourseID)
	if err != nil {
		return nil, model.ErrValidation
	}
	divisionID, err := uuid.Parse(req.DivisionID)
	if err != nil {
		return nil, model.ErrValidation
	}

	isActive := true
	if req.IsActive != nil {
		isActive = *req.IsActive
	}

	booklet := &model.Booklet{
		ID:           uuid.New(),
		CourseID:     courseID,
		DivisionID:   divisionID,
		Title:        req.Title,
		Description:  req.Description,
		CurrentPrice: req.Price,
		Stock:        req.Stock,
		ImageURL:     req.ImageURL,
		IsActive:     isActive,
	}
	if err := s.bookletRepo.Create(ctx, booklet); err != nil {
		return nil, err
	}
	return s.bookletRepo.FindByID(ctx, booklet.ID)
}

// UpdateBooklet updates an existing booklet.
func (s *CatalogService) UpdateBooklet(ctx context.Context, id uuid.UUID, req UpdateBookletRequest) (*model.Booklet, error) {
	if req.Price.IsNegative() {
		return nil, model.ErrNegativePrice
	}
	if req.Stock < 0 {
		return nil, model.ErrNegativeStock
	}

	courseID, err := uuid.Parse(req.CourseID)
	if err != nil {
		return nil, model.ErrValidation
	}
	divisionID, err := uuid.Parse(req.DivisionID)
	if err != nil {
		return nil, model.ErrValidation
	}

	booklet, err := s.bookletRepo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	booklet.CourseID = courseID
	booklet.DivisionID = divisionID
	booklet.Title = req.Title
	booklet.Description = req.Description
	booklet.CurrentPrice = req.Price
	booklet.Stock = req.Stock
	booklet.ImageURL = req.ImageURL
	if req.IsActive != nil {
		booklet.IsActive = *req.IsActive
	}

	if err := s.bookletRepo.Update(ctx, booklet); err != nil {
		return nil, err
	}
	return s.bookletRepo.FindByID(ctx, id)
}

// DeleteBooklet deletes a booklet by ID.
func (s *CatalogService) DeleteBooklet(ctx context.Context, id uuid.UUID) error {
	return s.bookletRepo.Delete(ctx, id)
}

File: backend/src/services/catalog.service.js
import { v4 as uuidv4 } from 'uuid';
import { prisma } from '../lib/prisma.js';

export class CatalogService {
  // Student-facing
  async listSchools() {
    return prisma.school.findMany({
      where: { isActive: true },
      include: {
        courses: {
          include: {
            course: {
              include: {
                divisions: {
                  where: { isActive: true },
                  orderBy: { name: 'asc' },
                },
              },
            },
          },
        },
      },
      orderBy: { name: 'asc' },
    });
  }

  async listCourses() {
    return prisma.course.findMany({
      where: { isActive: true },
      orderBy: { name: 'asc' },
    });
  }

  async listDivisionsByCourse(courseId) {
    const course = await prisma.course.findUnique({ where: { id: courseId } });
    if (!course) {
      const err = new Error('course not found');
      err.code = 'CAT_001';
      err.status = 404;
      throw err;
    }
    return prisma.division.findMany({
      where: { courseId, isActive: true },
      orderBy: { name: 'asc' },
    });
  }

  async listActiveBooklets({ courseId, divisionId, page = 1, limit = 20 }) {
    const where = { isActive: true, stock: { gt: 0 } };
    if (courseId) where.courseId = courseId;
    if (divisionId) where.divisionId = divisionId;

    const [booklets, total] = await Promise.all([
      prisma.booklet.findMany({
        where,
        orderBy: { title: 'asc' },
        skip: (page - 1) * limit,
        take: limit,
      }),
      prisma.booklet.count({ where }),
    ]);

    return { booklets, total, page, limit };
  }

  async getBooklet(id) {
    const booklet = await prisma.booklet.findUnique({ where: { id } });
    if (!booklet) {
      const err = new Error('booklet not found');
      err.code = 'CAT_003';
      err.status = 404;
      throw err;
    }
    return booklet;
  }

  // Admin — Courses
  async createCourse({ name, description }) {
    return prisma.course.create({
      data: { name, description, isActive: true },
    });
  }

  async updateCourse(id, { name, description, isActive }) {
    const course = await prisma.course.findUnique({ where: { id } });
    if (!course) {
      const err = new Error('course not found');
      err.code = 'CAT_001';
      err.status = 404;
      throw err;
    }
    return prisma.course.update({
      where: { id },
      data: {
        name,
        ...(description !== undefined && { description }),
        ...(isActive !== undefined && { isActive }),
      },
    });
  }

  async deleteCourse(id) {
    const result = await prisma.course.deleteMany({ where: { id } });
    if (result.count === 0) {
      const err = new Error('course not found');
      err.code = 'CAT_001';
      err.status = 404;
      throw err;
    }
  }

  // Admin — Divisions
  async listAllDivisions() {
    return prisma.division.findMany({ orderBy: [{ courseId: 'asc' }, { name: 'asc' }] });
  }

  async createDivision({ courseId, name }) {
    const course = await prisma.course.findUnique({ where: { id: courseId } });
    if (!course) {
      const err = new Error('course not found');
      err.code = 'CAT_001';
      err.status = 404;
      throw err;
    }
    return prisma.division.create({
      data: { courseId, name, isActive: true },
    });
  }

  async updateDivision(id, { courseId, name, isActive }) {
    const division = await prisma.division.findUnique({ where: { id } });
    if (!division) {
      const err = new Error('division not found');
      err.code = 'CAT_002';
      err.status = 404;
      throw err;
    }
    // Verify course exists
    const course = await prisma.course.findUnique({ where: { id: courseId } });
    if (!course) {
      const err = new Error('course not found');
      err.code = 'CAT_001';
      err.status = 404;
      throw err;
    }
    return prisma.division.update({
      where: { id },
      data: {
        courseId,
        name,
        ...(isActive !== undefined && { isActive }),
      },
    });
  }

  async deleteDivision(id) {
    const result = await prisma.division.deleteMany({ where: { id } });
    if (result.count === 0) {
      const err = new Error('division not found');
      err.code = 'CAT_002';
      err.status = 404;
      throw err;
    }
  }

  // Admin — Booklets
  async listAllBooklets({ courseId, divisionId, page = 1, limit = 20 }) {
    const where = {};
    if (courseId) where.courseId = courseId;
    if (divisionId) where.divisionId = divisionId;

    const [booklets, total] = await Promise.all([
      prisma.booklet.findMany({
        where,
        orderBy: { title: 'asc' },
        skip: (page - 1) * limit,
        take: limit,
        include: {
          course: {
            include: {
              schools: { include: { school: true } },
            },
          },
        },
      }),
      prisma.booklet.count({ where }),
    ]);

    return { booklets, total, page, limit };
  }

  async createBooklet({ courseId, divisionId, title, description, currentPrice, stock = 0, imageUrl, isActive = true }) {
    if (currentPrice < 0) {
      const err = new Error('price must be non-negative');
      err.code = 'CAT_005';
      err.status = 400;
      throw err;
    }
    if (stock < 0) {
      const err = new Error('stock must be non-negative');
      err.code = 'CAT_006';
      err.status = 400;
      throw err;
    }

    // Verify course and division exist
    const [course, division] = await Promise.all([
      prisma.course.findUnique({ where: { id: courseId } }),
      prisma.division.findUnique({ where: { id: divisionId } }),
    ]);
    if (!course || !division) {
      const err = new Error('invalid course_id or division_id');
      err.code = 'AUTH_004';
      err.status = 400;
      throw err;
    }

    return prisma.booklet.create({
      data: { courseId, divisionId, title, description, currentPrice, stock, imageUrl, isActive },
    });
  }

  async updateBooklet(id, { courseId, divisionId, title, description, currentPrice, stock, imageUrl, isActive }) {
    if (currentPrice !== undefined && currentPrice < 0) {
      const err = new Error('price must be non-negative');
      err.code = 'CAT_005';
      err.status = 400;
      throw err;
    }

    const booklet = await prisma.booklet.findUnique({ where: { id } });
    if (!booklet) {
      const err = new Error('booklet not found');
      err.code = 'CAT_003';
      err.status = 404;
      throw err;
    }

    // Verify course and division exist
    const [course, division] = await Promise.all([
      prisma.course.findUnique({ where: { id: courseId } }),
      prisma.division.findUnique({ where: { id: divisionId } }),
    ]);
    if (!course || !division) {
      const err = new Error('invalid course_id or division_id');
      err.code = 'AUTH_004';
      err.status = 400;
      throw err;
    }

    return prisma.booklet.update({
      where: { id },
      data: {
        courseId,
        divisionId,
        title,
        description,
        currentPrice,
        ...(stock !== undefined && { stock }),
        imageUrl,
        ...(isActive !== undefined && { isActive }),
      },
    });
  }

  async deleteBooklet(id) {
    const result = await prisma.booklet.deleteMany({ where: { id } });
    if (result.count === 0) {
      const err = new Error('booklet not found');
      err.code = 'CAT_003';
      err.status = 404;
      throw err;
    }
  }
}
